File: libs/go-pitwall/messaging/validate.go
package messaging

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/santhosh-tekuri/jsonschema/v6"
)

// Validator validates messages against the /contract JSON Schemas (blueprint
// §Messaging: validate every message in and out). It compiles the envelope schema,
// the heartbeat data schema, and an index of every event data schema keyed by its
// wire `type`, once at startup. A message that fails is logged + quarantined/rejected
// by the caller, never published nor applied (CLAUDE.md rule 5).
type Validator struct {
	envelope  *jsonschema.Schema
	heartbeat *jsonschema.Schema
	data      map[string]*jsonschema.Schema // wire type -> data-payload schema
}

// ResolveContractDir returns the directory holding the /contract tree. It honors
// CONTRACT_DIR (set explicitly in the container, where /contract is COPYed in) and
// otherwise walks up from the working directory to find it (dev/test).
func ResolveContractDir(explicit string) (string, error) {
	if explicit != "" {
		return explicit, nil
	}
	dir, err := os.Getwd()
	if err != nil {
		return "", err
	}
	for {
		cand := filepath.Join(dir, "contract", "schemas", "envelope.schema.json")
		if _, err := os.Stat(cand); err == nil {
			return filepath.Join(dir, "contract"), nil
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return "", fmt.Errorf("could not locate the /contract directory (set CONTRACT_DIR)")
		}
		dir = parent
	}
}

// NewValidator compiles the envelope schema, the heartbeat data schema, and every
// event data schema under contractDir. contractDir is the path to the /contract tree
// (see ResolveContractDir).
func NewValidator(contractDir string) (*Validator, error) {
	envSchema, err := compile(filepath.Join(contractDir, "schemas", "envelope.schema.json"))
	if err != nil {
		return nil, fmt.Errorf("compile envelope schema: %w", err)
	}
	hbSchema, err := compile(filepath.Join(contractDir, "schemas", "control", "control.heartbeat.v1.schema.json"))
	if err != nil {
		return nil, fmt.Errorf("compile heartbeat schema: %w", err)
	}
	data, err := indexDataSchemas(filepath.Join(contractDir, "schemas"))
	if err != nil {
		return nil, fmt.Errorf("index data schemas: %w", err)
	}
	return &Validator{envelope: envSchema, heartbeat: hbSchema, data: data}, nil
}

// schemaVersionSuffix matches the trailing version segment of a schema filename
// stem, e.g. "lap.recorded.v1" -> type "lap.recorded".
var schemaVersionSuffix = regexp.MustCompile(`\.v\d+$`)

// indexDataSchemas compiles every event data schema under schemasDir, keyed by its
// wire `type` (filename stem minus the trailing `.vN`). The envelope meta-schema is
// skipped (it is not a data payload).
func indexDataSchemas(schemasDir string) (map[string]*jsonschema.Schema, error) {
	out := map[string]*jsonschema.Schema{}
	err := filepath.WalkDir(schemasDir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() || !strings.HasSuffix(d.Name(), ".schema.json") || d.Name() == "envelope.schema.json" {
			return nil
		}
		stem := strings.TrimSuffix(d.Name(), ".schema.json")
		typ := schemaVersionSuffix.ReplaceAllString(stem, "")
		s, cerr := compile(path)
		if cerr != nil {
			return fmt.Errorf("compile %s: %w", path, cerr)
		}
		out[typ] = s
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

// ValidateHeartbeat validates the full envelope against the envelope schema and its
// data against the heartbeat data schema. A non-nil error means do not publish.
func (v *Validator) ValidateHeartbeat(env Envelope) error {
	inst, err := toInstance(env)
	if err != nil {
		return err
	}
	if err := v.envelope.Validate(inst); err != nil {
		return fmt.Errorf("envelope: %w", err)
	}
	m, ok := inst.(map[string]any)
	if !ok {
		return fmt.Errorf("envelope did not marshal to a JSON object")
	}
	if err := v.heartbeat.Validate(m["data"]); err != nil {
		return fmt.Errorf("data: %w", err)
	}
	return nil
}

// ValidateEnvelopeBytes validates a marshalled message: the full envelope against the
// envelope schema, then its data against the schema registered for the envelope's
// `type`. It fails closed — an unknown type (no registered data schema) is an error,
// because a service must never publish nor apply an event it cannot validate. A
// non-nil error means: do not publish/process; quarantine or reject the message.
func (v *Validator) ValidateEnvelopeBytes(payload []byte) error {
	inst, err := jsonschema.UnmarshalJSON(bytes.NewReader(payload))
	if err != nil {
		return fmt.Errorf("payload is not valid JSON: %w", err)
	}
	if err := v.envelope.Validate(inst); err != nil {
		return fmt.Errorf("envelope: %w", err)
	}
	m, ok := inst.(map[string]any)
	if !ok {
		return fmt.Errorf("envelope did not marshal to a JSON object")
	}
	typ, _ := m["type"].(string)
	ds, ok := v.data[typ]
	if !ok {
		return fmt.Errorf("no /contract data schema for type %q", typ)
	}
	if err := ds.Validate(m["data"]); err != nil {
		return fmt.Errorf("data: %w", err)
	}
	return nil
}

func compile(path string) (*jsonschema.Schema, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	doc, err := jsonschema.UnmarshalJSON(bytes.NewReader(raw))
	if err != nil {
		return nil, err
	}
	c := jsonschema.NewCompiler()
	const id = "schema.json"
	if err := c.AddResource(id, doc); err != nil {
		return nil, err
	}
	return c.Compile(id)
}

// toInstance converts a Go value into the native JSON representation
// (map[string]any / float64 / ...) that the validator expects.
func toInstance(v any) (any, error) {
	b, err := json.Marshal(v)
	if err != nil {
		return nil, err
	}
	return jsonschema.UnmarshalJSON(bytes.NewReader(b))
}

File: libs/py-pitwall/pitwall/validate.py
"""The /contract JSON-Schema validator wrapper (blueprint §Messaging: validate every
message in and out). Compiles the envelope schema, the heartbeat data schema, and an
index of every event data schema keyed by its wire `type`, once at construction. A
message that fails is logged + quarantined/rejected by the CALLER, never published nor
applied (CLAUDE.md rule 5) — this module only raises; callers decide the sad-path.

Deliberately independent of the /contract-generated Pydantic models (contract/codegen/
python) — codegen and validation are separate concerns here exactly as in the Go
library (architecture §Schema-as-Source & Codegen: "the validator is a thin
per-language wrapper over a standard JSON-Schema library").
"""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

import jsonschema
from jsonschema.validators import Draft202012Validator

_VERSION_SUFFIX = re.compile(r"\.v\d+$")


def resolve_contract_dir(explicit: str | None) -> str:
    """Returns the directory holding the /contract tree. Honors an explicit path (set
    in the container, where /contract is COPYed in) and otherwise walks up from the
    working directory to find it (dev/test) — mirrors Go's ResolveContractDir."""
    if explicit:
        return explicit
    d = Path.cwd()
    while True:
        candidate = d / "contract" / "schemas" / "envelope.schema.json"
        if candidate.is_file():
            return str(d / "contract")
        if d.parent == d:
            raise RuntimeError("could not locate the /contract directory (set CONTRACT_DIR)")
        d = d.parent


class ContractValidationError(Exception):
    """Raised when a message fails /contract validation (envelope or data shape) or
    names a `type` with no registered data schema (fail closed — a service must never
    publish nor apply an event it cannot validate)."""


def _compile(path: Path) -> Draft202012Validator:
    with path.open("r", encoding="utf-8") as f:
        schema = json.load(f)
    validator_cls = jsonschema.validators.validator_for(schema, default=Draft202012Validator)
    validator_cls.check_schema(schema)
    return validator_cls(schema)


def _index_data_schemas(schemas_dir: Path) -> dict[str, Draft202012Validator]:
    out: dict[str, Draft202012Validator] = {}
    for path in schemas_dir.rglob("*.schema.json"):
        if path.name == "envelope.schema.json":
            continue
        stem = path.name.removesuffix(".schema.json")
        wire_type = _VERSION_SUFFIX.sub("", stem)
        out[wire_type] = _compile(path)
    return out


class Validator:
    """Validates messages against the /contract JSON Schemas."""

    def __init__(self, contract_dir: str):
        schemas_dir = Path(contract_dir) / "schemas"
        self._envelope = _compile(schemas_dir / "envelope.schema.json")
        self._heartbeat = _compile(schemas_dir / "control" / "control.heartbeat.v1.schema.json")
        self._data = _index_data_schemas(schemas_dir)

    def validate_envelope_bytes(self, payload: bytes) -> None:
        """Validates a marshalled message: the full envelope against the envelope
        schema, then its data against the schema registered for the envelope's `type`.
        Fails closed on an unknown type. Raises ContractValidationError (or a
        jsonschema.ValidationError for the envelope/data shape itself) — never returns
        a value; a clean return means valid."""
        try:
            instance = json.loads(payload)
        except json.JSONDecodeError as e:
            raise ContractValidationError(f"payload is not valid JSON: {e}") from e

        self._envelope.validate(instance)

        wire_type = instance.get("type")
        data_validator = self._data.get(wire_type)
        if data_validator is None:
            raise ContractValidationError(f"no /contract data schema for type {wire_type!r}")
        data_validator.validate(instance.get("data"))

    def validate_heartbeat(self, envelope: dict[str, Any]) -> None:
        """Validates a full envelope dict against the envelope schema and its data
        against the heartbeat data schema specifically (used when constructing a
        heartbeat before publish, mirroring Go's ValidateHeartbeat)."""
        self._envelope.validate(envelope)
        self._heartbeat.validate(envelope.get("data"))
